Source/WebCore/platform/text/TextFlags.cpp

```diff
@@ -58,15 +60,69 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
     return ts;
 }
 
+template <typename T> using BinOp = std::function<T(T, T)>;
+
+template <typename T>
+T join(const std::vector<T>& vec, const BinOp<T>& binOp)
+{
+    if (vec.empty())
+        return { };
+
+    if (vec.size() == 1)
+        return vec[0];
+    
+    auto acc = vec[0];
+    for (size_t i = 1 ; i < vec.size() ; i++)
+        acc = binOp(acc, vec[i]);
+
+    return acc;
+}
+
 WTF::TextStream& operator<<(TextStream& ts, FontVariantAlternates alternates)
 {
-    switch (alternates) {
-    case FontVariantAlternates::Normal: ts << "normal"; break;
-    case FontVariantAlternates::HistoricalForms: ts << "historical-forms"; break;
+    if (alternates.isNormal())
+        ts << "normal";
+    else {
+        auto values = alternates.values();
+        std::vector<String> result;
+        if (values.stylistic)
+            result.push_back("stylistic(" + *values.stylistic + ")");
+        if (values.historicalForms)
+            result.push_back("historical-forms"_s);
+        if (values.styleset)
+            result.push_back("styleset(" + *values.styleset + ")");
+        if (values.characterVariant)
+            result.push_back("character-variant(" + *values.characterVariant + ")");
+        if (values.swash)
+            result.push_back("swash(" + *values.swash + ")");
+        if (values.ornaments)
+            result.push_back("ornaments(" + *values.ornaments + ")");
+        if (values.annotation)
+            result.push_back("annotation(" + *values.annotation + ")");
```

Comment:
I kind of feel this is a layering violation, where the serialization rules here are defined in CSS, for CSS. This code is in platform/ and shouldn't know anything about CSS serialization rules. I think this function should be hoisted up to `CSSFontVariantAlternatesValue`.

Reply:
It follows the CSS serialization rule because it's a convenient serialisation, but it's not mandatory indeed (because it's in platform).

I could move it to the css/ folder, but I would need roughly the exact same code here to implement the `<< stream` operator. WDYT ?

Reply:
I've added a comment in the`CSSFontVariantAlternatesValue::customCSSText()`.

Reply:
ok